### ISSUE TITLE
People Merging

### ISSUE THREAD

**Issue by nktm85** [photoprism/photoprism#3292]
I'm not sure how to best describe this, but the face detection is not allowing me to separate some faces that Photo Prism thinks is 1 person, when in fact they are two different people.  
I added a name to a New faces detection, for explanation [Person A], and upon looking at all the pictures I noticed a few pictures that it auto detected another person [Person B].  I went through the entire faces and removed [Person B's] face from [Person's A auto detection.  I then went back into New faces, and see [Person B's] face that I deleted, and properly labelled to [Person B's] name.  Upon doing so, ALL of [Person A's] pictures that were ALREADY tagged to [Person A's] proper name got applied to [Person B's] name.  
It goes both ways, any time I label [Person A] that Photo Prism tagged as [Person B] and ALL the photos are now under [Person A's] name.  
The people that Photo Prism is detecting aren't really that similar, but it's the fact that I cannot separate the names makes it very difficult to tag the right person.  I'm not sure how to resolve this, and I'm hoping this won't happen for other people in my library as I do have a good number of twins in my family (interestingly this issue came from non-twins).

Is there a way to train the face detection?  I know Google used to ask if two faces are the same persons.

**Comment by hieronymousch** [comment]
I have this issue as well for my twins. There should be a parameter for person8to disable the merge or remove I think

**Comment by hieronymousch** [comment]
Seems a duplicate of https://github.com/photoprism/photoprism/issues/1497#issuecomment-1427820237

**Comment by kevincox** [comment]
I think this is different. This isn't a bug with face detection, it is a problem with managing "people". I've run into this myself and it can be disastrous.

# Case 1

I was manually tagging faces and I accidentally marked "Alice" as "Bob". Photoprism then marked all existing Alice photos as Bob. Now these two people can't be separated. The first time it wasn't so bad because I only had 1 Bob photo. So I renamed Bob to Alice, then manually tagged the 1 real Bob photo as "Bob 2" (for some reason calling it Bob would just merge the two again for a few days)

# Case 2

I had thousands of pictures of both Alice and Bob. I accidentally marked one Alice photo as Bob. This merged both people together. Now I need to spend hours manually retagging all of these faces.

---

This behaviour is just outright wrong. If I tag one photo as Bob that shouldn't affect existing Alice photos. Not only that but it means missing a single click target can ruin your whole photo library.

**Comment by kevincox** [comment]
This completely messed up my library. I tried manually separating the two people but at some point they were collapsed back together (except for 12 photos). Back to square one. Manually tagged faces should **never** be changed except by explicit retagging.

I decided that the best option was to nuke the mess of merged faces and start over again. I'm going to again need to spend hours manually retagging thousands faces but hopefully they won't collapse back together on their own.

The approach that I took was this **USE AT YOUR OWN RISK**.

1. Connect to the sqlite database.
2. Run `SELECT subj_name, subj_uid FROM subjects;` to identify the relevant `subj_uid`.
3. Use the `.schema` command to identify all tables that have a `subj_uid` column.
4. Use `DELETE FROM {table} WHERE subj_uid = '{id}'` command for each of those tables.

Then I triggered a complete re-index. I'll update this comment with any problems I run into with this approach.

**Comment by kevincox** [comment]
So that strategy mostly worked. The only real issue (other than huge amounts of manual retagging) is that some of these faces got grouped in with other people. Often this was easy to fix but there were a few cases where I ran into this issue where trying to split out faces ended up with them in the same group. This happened for both of the people that I ended up clearing.

1. Alice (thousands of photos) got merged into Bob (1 photo)
2. Rename Bob to Alice (as most of the photos are of Alice).
3. Remove the one Bob photo and label it as Bob (a "new" person).
4. Now all Alice photos are called "Bob" again.

This also happened with a different 2 people where the "Bob" there had 5 photos.

I went to explore the DB to figure out why this was happening but ended up not finding any instances of "Bob" after step 2. However when I tried renaming again it worked. So IDK what changed. I did rename the photos in a different order, so maybe there is only one "core photo" or something that will trigger this if you do it first? But updating that photo last didn't trigger a full re-merge 🤷 So it may be hard to reproduce, but there is **definitely** a major and destructive bug here.

**Comment by graciousgrey** [comment]
@kevincox Thank you for trying to debug this. Once we have completed the work on OpenID Connect support, we will address the face recognition related issues. We might then be able to take a look at a database dump from you to see if we can find the underlying problem.

**Comment by Melsana** [comment]
I'm running into this bug too.  It is quite frustrating.

**Comment by SinTan1729** [comment]
I'm running into this issue trying to distinguish between my twin nieces. I can easily distinguish between them, but PhotoPrism collapses both persons into one. I think PhotoPrism should distinguish between automatically detected and manually added names, and never mess with manually added names while doing automatic person detection.

(Also, it's weird that it detects both faces in a single picture to be the same person.)

**Comment by lastzero** [comment]
If there's a mirror or it's a collage, the same face can appear multiple times in the same image.

**Comment by marcotama** [comment]
My gallery is 80k+ photos worth. These are the photos of an average person through about 20 years of smartphone pics, not a photographer's collection. Not only the AI facial recognition is so weak (separate issue, see #4669); to make matters worse:
1. the workflow for managing people in the gallery is severely lacking (e.g. how do I rename a person? can I change the subject id? how do I batch-mark several faces to a new person? why instead of "ejecting" a person from a photo don't we start a workflow where I can tell you who that person really is (new or existing) and you propose me a carousel of likely pics that need updating?)
2. the software is extremely slow when it comes to selecting or updating a person for a face in a photo. I have had to wait even 30 seconds after hitting Enter. How am I supposed to get through a gallery of my size?

**Comment by photoprismci** [comment]
Thanks for all the reports here — an update now that the latest release is out.

Merging people when naming a face is fixed (#5764). A name field is read-only while a face is assigned, so this only happened when the browser's view was out of date, for example with the picture open in a second tab: the server then read the submitted name as a rename of the assigned person and combined the two. It no longer relies on what the browser believes — a name that belongs to an existing person links the face to that person and leaves the other one unchanged. Combining two people stays an explicit action on the People page. It does not undo a merge that already happened; the recovery options are described under [Inconsistent Face Assignments](https://docs.photoprism.app/known-issues/#inconsistent-face-assignments).

Face recognition has been improved by using a better trained model (#4669). That should also help where two people who look alike previously ended up in the same cluster and naming it named both. Grouping faces by similarity can never be perfect, though, so some manual correction will always be needed.

Renaming and performance. Renaming a person is possible and does not involve the subject ID: People → Recognized → click the name → type the new one → Save ([user guide](https://docs.photoprism.app/user-guide/organize/people/#rename-people)). This release also improves the performance of assigning a person to a face. Further improvements, such as naming multiple new faces at once (#1679), are tracked in separate issues.
